app/src/main/java/de/westnordost/streetcomplete/quests/opening_hours/OpeningHoursTagParser.kt

```diff
@@ -0,0 +1,463 @@
+package de.westnordost.streetcomplete.quests.opening_hours
+
+import ch.poole.openinghoursparser.*
+import de.westnordost.streetcomplete.quests.opening_hours.adapter.OpeningMonthsRow
+import de.westnordost.streetcomplete.quests.opening_hours.adapter.OpeningWeekdaysRow
+import de.westnordost.streetcomplete.quests.opening_hours.model.CircularSection
+import de.westnordost.streetcomplete.quests.opening_hours.model.OpeningMonths
+import de.westnordost.streetcomplete.quests.opening_hours.model.TimeRange
+import de.westnordost.streetcomplete.quests.opening_hours.model.Weekdays
+import java.io.ByteArrayInputStream
+
+object OpeningHoursTagParser {
```

Comment:
This should be a normal class like any other, injected the same like other dependencies of classes because otherwise it is an implicit dependency, makes unit testing of classes that use this class harder and obscures the real dependencies of a class.

Reply:
injection dependency is part where my understanding is weakest of all what is happening in this app (and I probably should change that), but `OpeningHoursTagParser` is used a bag of functions, it is not used to create objects. Like `DateUtil`.

I have no idea how dependency injection would work here. Drop static functions and start creating and passing `OpeningHoursTagParser` object?

Is there any other class that is handled in a similar way?

Reply:
Just replace `object` with `class` in line 12, then at every place where you use  `OpeningHoursTagParser`, declare it as a constructor parameter. In the test, of course, you create it yourself.

See for example `AddPlaceName` which depends on (a Future of) the `FeatureDictionary`

Reply:
There is no magic in dependency injection. It basically just means that every class that uses another class should get that classed passed in and not create it itself.

```
class Car(val motor: Motor)
```
^- Car depends on a motor, this dependency is injected by whoever instantiated the car.

Reply:
I modified the class in https://github.com/westnordost/StreetComplete/pull/1676/commits/3ddd5b4bf22b2460b33ce96d58f5d4388724444f, it should now use DI and it is also still working.

> There is no magic in dependency injection. It basically just means that every class that uses another class should get that class passed in and not create it itself.

Oh, basic idea for dependency injection is clear for me. It is so simple that I independently noticed it as a good idea during refactoring one of my projects (long time ago, before I even learned about terms like "dependency injection" or "refactoring" and before I was using git).

But in practice Dagger is for my highly confusing. Whenever I learn about @Module vs @Component difference I will forget which one was doing what within 20 minutes.

For more specific example - why I added https://github.com/westnordost/StreetComplete/pull/1676/files#diff-4d3f82a0d1eeea6b01ea730c337174d7 ?

ApplicationComponent is a `@Component` so it will be (somehow) used by Dagger to generate code used by ApplicationModule. 

But I admit that adding `void inject(ResurveyOpeningHoursForm resurveyOpeningHoursForm);` in this PR was basically a magic incantation to exercise compilation error without understanding why I am doing this.

I again run went through a bit of official Dagger docs, code handling Dagger, code generated by Dagger, and some tutorial found somewhere. It seems that I understand a bit more now, but I am not entirely sure about this.

Reply:
To be honest, I never got so deep into Dagger myself, I only worked myself into this as far as I needed to to have everything working as I want it to be. It could be that the module / component code in this project is not best practice.

The `void inject(ResurveyOpeningHoursForm resurveyOpeningHoursForm);` in the interface of the ApplicationComponent is just to tell Dagger that it should also generate a method (this method) which should inject dependencies into an already created object when invoked. This is necessary, because in Android development, we can not simply create new activities and new fragment objects ourselves using the constructor. The Android system creates these objects for us, so the "normal" way we'd go for dependency injection is not available. The workaround is to have the class inject its dependencies after construction itself, by calling the method of the application component itself.
In my opinion, this is only a workaround for this situation and is less than ideal. For example, this makes unit testing of that fragment more difficult again because the dependencies are indeed not really injected but the fragment injects them itself by telling Dagger to inject them.